src/par/intrinsics/deinterleave.rs

```diff
@@ -0,0 +1,127 @@
+use wide::u32x8;
+
+#[inline(always)]
+#[cfg(all(
+    any(target_arch = "x86", target_arch = "x86_64"),
+    target_feature = "avx2"
+))]
+unsafe fn deinterleave_avx(a: u32x8, b: u32x8) -> (u32x8, u32x8) {
+    #[cfg(target_arch = "x86")]
+    use core::arch::x86::{__m256, __m256d, _mm256_permute4x64_pd, _mm256_shuffle_ps};
+    #[cfg(target_arch = "x86_64")]
+    use core::arch::x86_64::{__m256, __m256d, _mm256_permute4x64_pd, _mm256_shuffle_ps};
+    use core::mem::transmute;
+
+    const SHUFFLE_EVEN: i32 = if cfg!(target_endian = "little") {
+        0b10_00_10_00
+    } else {
+        0b00_10_00_10
+    };
+    const SHUFFLE_ODD: i32 = if cfg!(target_endian = "little") {
+        0b11_01_11_01
+    } else {
+        0b01_11_01_11
+    };
+    const SWAP_MIDDLE: i32 = if cfg!(target_endian = "little") {
+        0b11_01_10_00
+    } else {
+        0b00_10_01_11
+    };
+
+    let a: __m256 = transmute(a);
+    let b: __m256 = transmute(b);
+    let abab_even: __m256d = transmute(_mm256_shuffle_ps(a, b, SHUFFLE_EVEN));
+    let abab_odd: __m256d = transmute(_mm256_shuffle_ps(a, b, SHUFFLE_ODD));
+    let ab_even = _mm256_permute4x64_pd(abab_even, SWAP_MIDDLE);
+    let ab_odd = _mm256_permute4x64_pd(abab_odd, SWAP_MIDDLE);
+    (transmute(ab_even), transmute(ab_odd))
+}
+
+#[inline(always)]
+#[cfg(all(target_arch = "aarch64", target_feature = "neon"))]
+unsafe fn deinterleave_neon(a: u32x8, b: u32x8) -> (u32x8, u32x8) {
+    #[cfg(target_arch = "aarch64")]
+    use core::arch::aarch64::{uint32x4_t, vuzp1q_u32, vuzp2q_u32};
+    use core::mem::transmute;
+
+    let (a1, a2): (uint32x4_t, uint32x4_t) = transmute(a);
+    let (b1, b2): (uint32x4_t, uint32x4_t) = transmute(b);
+    let a_even = vuzp1q_u32(a1, a2);
+    let a_odd = vuzp2q_u32(a1, a2);
+    let b_even = vuzp1q_u32(b1, b2);
+    let b_odd = vuzp2q_u32(b1, b2);
+    let ab_even: u32x8 = transmute((a_even, b_even));
+    let ab_odd: u32x8 = transmute((a_odd, b_odd));
+    (ab_even, ab_odd)
+}
+
+#[inline(always)]
+#[deprecated(
```

Comment:
I'd rather that we put a conditional deprecation message on the entrypoints of the library.

Reply:
Okay, how would you do that?